src/Passes/Source/RuleTransformationPass/RulePass.hpp

```diff
@@ -0,0 +1,245 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Logging/ILogger.hpp"
+#include "Profile/Profile.hpp"
+#include "RuleTransformationPass/RuleTransformationPassConfiguration.hpp"
+#include "Rules/RuleSet.hpp"
+#include "Types/Types.hpp"
+
+#include "Llvm/Llvm.hpp"
+
+#include <functional>
+#include <unordered_map>
+#include <vector>
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    /// This class applies a set of transformation rules to the IR to transform it into a new IR. The
+    /// rules are added using the RuleSet class which allows the developer to create one or more rules
+    /// on how to transform the IR.
+    ///
+    ///
+    /// The module executes the following steps:
+    ///
+    ///
+    ///
+    ///           ┌─────────────────┐
+    ///           │  Apply profile  │
+    ///           └─────────────────┘
+    ///                    │
+    ///                    │
+    ///                    │
+    ///                    │
+    ///                    │                ┌───────────────────────────────┐
+    ///                    │                │                               │
+    ///                    ├───────────────▶│   Copy and expand functions   │──┐
+    ///                    │     clone      │                               │  │
+    ///                    │   functions?   └───────────────────────────────┘  │
```

Comment:
FYI (confusion): I don't know what "clone functions", "Copy functions", and "expand functions" mean.
(with "Copy functions" you don't mean Copy Constructor and Copy-Assignment Operator, do you?)

Intuitively I assume that you might mean 
"Inline the functions?" (instead of "clone functions?"),
"Inline the functions" (instead of "Copy and expand functions").

Reply:
Not quite: It literally makes a copy of the function body replacing any compile-time arguments. I've added a text to clarify this.